{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0eba0072",
   "metadata": {},
   "source": [
    "# RNA-seq Data Analysis Pipeline\n",
    "## Date: Tuesday 26/11/2024\n",
    "\n",
    "This notebook contains the complete workflow for RNA-seq data analysis, including quality control, preprocessing, and mapping steps. The notebook is designed to be self-explanatory and includes detailed documentation for future reference.\n",
    "\n",
    "## Overview\n",
    "This pipeline processes paired-end RNA-seq data from mouse samples. We will:\n",
    "1. Check the quality of raw reads using FastQC\n",
    "2. Preprocess reads using fastp\n",
    "3. Verify the quality of processed reads\n",
    "4. Download and index the reference genome\n",
    "5. Map reads to the reference genome using STAR\n",
    "6. Generate comprehensive quality reports with MultiQC\n",
    "\n",
    "## Data Organization\n",
    "- Raw data location: `/srv/data/meg-m2-rnaseq/Data/fastq/raw/`\n",
    "- Genome annotation: `/srv/data/meg-m2-rnaseq/Genomes/Mmu/GRCm39/extracted/`\n",
    "- Results will be organized in tool-specific subfolders:\n",
    "  - FastQC results: `Results/fastqc/`\n",
    "  - fastp results: `Results/fastp/`\n",
    "  - STAR mapping results: `Results/star/`\n",
    "  - MultiQC results: `Results/multiqc/`\n",
    "\n",
    "## Required Software\n",
    "- FastQC v0.12.1\n",
    "- MultiQC v1.13\n",
    "- fastp v0.23.1\n",
    "- STAR v2.7.11a\n",
    "- samtools v1.18\n",
    "\n",
    "## Important Notes\n",
    "- This notebook demonstrates the workflow on the first two samples\n",
    "- Commands for processing all samples are provided as commented code\n",
    "- Results are organized in tool-specific directories\n",
    "- The pipeline is designed for mouse RNA-seq data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128709a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "# Create Results directory structure\n",
    "mkdir -p Results/{fastqc,fastp,star,multiqc}\n",
    "\n",
    "# Display the created directory structure\n",
    "tree Results/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "432185cb",
   "metadata": {},
   "source": [
    "## Reference Genome Download and Indexing\n",
    "\n",
    "### Important Notes\n",
    "- The mouse reference genome (GRCm39) will be downloaded from Ensembl\n",
    "- We'll use STAR for indexing the genome\n",
    "- This step requires significant computational resources\n",
    "- The index will be used for all subsequent mapping steps\n",
    "\n",
    "### Key Parameters for STAR Indexing\n",
    "- `--runMode genomeGenerate`: Tells STAR to create an index\n",
    "- `--genomeDir`: Directory where the index will be stored\n",
    "- `--genomeFastaFiles`: Path to the reference genome FASTA\n",
    "- `--sjdbGTFfile`: Path to the GTF annotation file\n",
    "- `--runThreadN`: Number of threads (we'll use 10)\n",
    "\n",
    "### Documentation Links\n",
    "- [Ensembl Mouse Genome](https://www.ensembl.org/Mus_musculus/Info/Index)\n",
    "- [STAR Manual - Genome Generation](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf#page=7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20d15348",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download and prepare reference genome\n",
    "mkdir -p reference\n",
    "cd reference\n",
    "\n",
    "# Download mouse reference genome from Ensembl\n",
    "wget https://ftp.ensembl.org/pub/release-109/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\n",
    "gunzip Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\n",
    "\n",
    "# Create STAR index\n",
    "mkdir -p star_index\n",
    "STAR --runMode genomeGenerate \\\n",
    "     --genomeDir star_index \\\n",
    "     --genomeFastaFiles Mus_musculus.GRCm39.dna.primary_assembly.fa \\\n",
    "     --sjdbGTFfile /srv/data/meg-m2-rnaseq/Genomes/Mmu/GRCm39/extracted/mouse\\ genome_annotation-M35.gtf \\\n",
    "     --runThreadN 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23c8b9c4",
   "metadata": {},
   "source": [
    "## Quality Control with FastQC\n",
    "\n",
    "### Overview\n",
    "FastQC performs quality control checks on raw sequence data. We'll analyze:\n",
    "- Basic Statistics\n",
    "- Per base sequence quality\n",
    "- Per sequence quality scores\n",
    "- Sequence duplication levels\n",
    "- Overrepresented sequences\n",
    "\n",
    "### Important Parameters\n",
    "- `-o`: Output directory\n",
    "- `-t`: Number of threads\n",
    "- `--noextract`: Don't extract zip files\n",
    "\n",
    "### Documentation\n",
    "- [FastQC Documentation](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)\n",
    "- [Example Reports](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60f3dbc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run FastQC on first two samples\n",
    "DATA_DIR=\"/srv/data/meg-m2-rnaseq/Data/fastq/raw\"\n",
    "fastqc -o Results/fastqc -t 10 \\\n",
    "    ${DATA_DIR}/sample1_R1.fastq.gz \\\n",
    "    ${DATA_DIR}/sample1_R2.fastq.gz \\\n",
    "    ${DATA_DIR}/sample2_R1.fastq.gz \\\n",
    "    ${DATA_DIR}/sample2_R2.fastq.gz\n",
    "\n",
    "# Command for all samples (shown but not executed)\n",
    ": '\n",
    "fastqc -o Results/fastqc -t 10 ${DATA_DIR}/*.fastq.gz\n",
    "'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15b4fcbb",
   "metadata": {},
   "source": [
    "## Read Preprocessing with fastp\n",
    "\n",
    "### Overview\n",
    "fastp performs quality control and preprocessing:\n",
    "- Adapter trimming\n",
    "- Quality filtering\n",
    "- Base correction\n",
    "- Quality reporting\n",
    "\n",
    "### Key Parameters\n",
    "- `--in1/--in2`: Input paired-end files\n",
    "- `--out1/--out2`: Output paired-end files\n",
    "- `--html`: HTML report output\n",
    "- `--json`: JSON report output\n",
    "- `--thread`: Number of threads\n",
    "\n",
    "### Documentation\n",
    "- [fastp GitHub](https://github.com/OpenGene/fastp)\n",
    "- [Parameters Guide](https://github.com/OpenGene/fastp#all-options)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78bb6bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process first two samples with fastp\n",
    "DATA_DIR=\"/srv/data/meg-m2-rnaseq/Data/fastq/raw\"\n",
    "for sample in sample1 sample2; do\n",
    "    fastp --in1 ${DATA_DIR}/${sample}_R1.fastq.gz \\\n",
    "          --in2 ${DATA_DIR}/${sample}_R2.fastq.gz \\\n",
    "          --out1 Results/fastp/${sample}_R1_cleaned.fastq.gz \\\n",
    "          --out2 Results/fastp/${sample}_R2_cleaned.fastq.gz \\\n",
    "          --html Results/fastp/${sample}_report.html \\\n",
    "          --json Results/fastp/${sample}_report.json \\\n",
    "          --thread 10\n",
    "done\n",
    "\n",
    "# Command for all samples (shown but not executed)\n",
    ": '\n",
    "for sample in sample*; do\n",
    "    fastp --in1 ${DATA_DIR}/${sample}_R1.fastq.gz \\\n",
    "          --in2 ${DATA_DIR}/${sample}_R2.fastq.gz \\\n",
    "          --out1 Results/fastp/${sample}_R1_cleaned.fastq.gz \\\n",
    "          --out2 Results/fastp/${sample}_R2_cleaned.fastq.gz \\\n",
    "          --html Results/fastp/${sample}_report.html \\\n",
    "          --json Results/fastp/${sample}_report.json \\\n",
    "          --thread 10\n",
    "done\n",
    "'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3200802f",
   "metadata": {},
   "source": [
    "## Read Mapping with STAR\n",
    "\n",
    "### Overview\n",
    "STAR (Spliced Transcripts Alignment to a Reference) performs RNA-seq alignment:\n",
    "- Handles splice junctions\n",
    "- Supports paired-end reads\n",
    "- Provides mapping statistics\n",
    "\n",
    "### Important Parameters\n",
    "- `--runThreadN`: Number of threads\n",
    "- `--genomeDir`: Path to genome index\n",
    "- `--readFilesIn`: Input files (R1 R2 for paired-end)\n",
    "- `--outFileNamePrefix`: Prefix for output files\n",
    "- `--outSAMtype`: Output format specification\n",
    "\n",
    "### Documentation\n",
    "- [STAR Manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)\n",
    "- [RNA-STAR GitHub](https://github.com/alexdobin/STAR)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58d0b324",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map first two samples with STAR\n",
    "for sample in sample1 sample2; do\n",
    "    STAR --runThreadN 10 \\\n",
    "         --genomeDir reference/star_index \\\n",
    "         --readFilesIn Results/fastp/${sample}_R1_cleaned.fastq.gz \\\n",
    "                      Results/fastp/${sample}_R2_cleaned.fastq.gz \\\n",
    "         --readFilesCommand zcat \\\n",
    "         --outFileNamePrefix Results/star/${sample}_ \\\n",
    "         --outSAMtype BAM SortedByCoordinate \\\n",
    "         --outBAMsortingThreadN 10\n",
    "done\n",
    "\n",
    "# Index BAM files\n",
    "for sample in sample1 sample2; do\n",
    "    samtools index Results/star/${sample}_Aligned.sortedByCoord.out.bam\n",
    "done\n",
    "\n",
    "# Commands for all samples (shown but not executed)\n",
    ": '\n",
    "for sample in sample*; do\n",
    "    STAR --runThreadN 10 \\\n",
    "         --genomeDir reference/star_index \\\n",
    "         --readFilesIn Results/fastp/${sample}_R1_cleaned.fastq.gz \\\n",
    "                      Results/fastp/${sample}_R2_cleaned.fastq.gz \\\n",
    "         --readFilesCommand zcat \\\n",
    "         --outFileNamePrefix Results/star/${sample}_ \\\n",
    "         --outSAMtype BAM SortedByCoordinate \\\n",
    "         --outBAMsortingThreadN 10\n",
    "\n",
    "    samtools index Results/star/${sample}_Aligned.sortedByCoord.out.bam\n",
    "done\n",
    "'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "237e8089",
   "metadata": {},
   "source": [
    "## Quality Report Aggregation with MultiQC\n",
    "\n",
    "### Overview\n",
    "MultiQC aggregates results from multiple bioinformatics tools:\n",
    "- FastQC reports\n",
    "- fastp reports\n",
    "- STAR alignment logs\n",
    "- Creates a single comprehensive report\n",
    "\n",
    "### Important Parameters\n",
    "- `-o`: Output directory\n",
    "- `-f`: Force overwrite existing reports\n",
    "\n",
    "### Documentation\n",
    "- [MultiQC Documentation](https://multiqc.info/)\n",
    "- [Available Modules](https://multiqc.info/docs/#available-modules)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c95684",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run MultiQC on all results\n",
    "multiqc -o Results/multiqc Results/"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
